secrets/vault/src/main/kotlin/VaultSecretsProvider.kt

```diff
@@ -153,27 +155,36 @@ class VaultSecretsProvider(
      * bearer-authentication offered by KTor cannot be used here.
      */
     private fun configureTokenHandling(client: HttpClient) {
-        val token = runBlocking { AtomicReference(fetchToken(client)) }
+        var token: String? = null
         val mutex = Mutex()
 
         client.plugin(HttpSend).intercept { request ->
-            val currentToken = token.get()
-            request.header(TOKEN_HEADER, currentToken)
-            val call = execute(request)
-
-            if (call.response.status == HttpStatusCode.Forbidden) {
-                // Synchronize, since theoretically there could be multiple callers at the same time.
-                mutex.withLock {
-                    // Check that no other caller has already renewed the token.
-                    if (token.get() == currentToken) {
-                        token.set(fetchToken(client))
-                    }
-                }
-
-                request.headers[TOKEN_HEADER] = token.get()
+            if (request.url.encodedPath == LOGIN_PATH) {
+                // This is the login request; execute it directly.
                 execute(request)
             } else {
-                call
+                val currentToken = mutex.withLock { token }
```

Comment:
IIUC the reason to use the lock here is to wait if there is a running token request. Could you add a code comment to explain that?

Reply:
Well, basic protection against concurrent access to shared data. I have added a comment to the `token` variable.